### PR TITLE
[Test] Add a more direct test of multi-gpu training working

### PR DESCRIPTION
**Pull request by maximsch2**
Haven't tested locally, hoping to see CI run on it to confirm test is passing. In general, the current test doesn't actually check that the model returned from training is changed, which is a common bug with DDP training.

### CODE REVIEW

**Comment by williamFalcon** [comment]
great! yeah, we actually do test that. i need to find the test but basically it trains a model, loads a checkpoint and verifies an accuracy. 

ddp was changed last week and before that the actual trained model did not change because it was in a different process. So, now it’s possible :)  And thus this test can exist now. :)

**Comment by williamFalcon** [comment]
not using interactive jupyter notebooks, but via actual scripts on clusters. ddp breaks on interactive environments for that reason (it doesn't now).

**Comment by maximsch2** [comment]
@williamFalcon , 

Here is the thing I'm trying to run locally (and it hangs on the DDP part) that I'm trying to repro on your end by adding the test. I wonder if you see any differences between what I'm trying to run vs what gets run in that test?

```
#!/usr/bin/env python3
import pytorch_lightning as pl
import torch
from torch import nn
from torch.utils.data import DataLoader, TensorDataset


class LogisticRegressionClassifier(pl.LightningModule):
    def __init__(self, dataset):
        super().__init__()
        self.criterion = nn.BCELoss()
        self.dataset = dataset
        self.linear = nn.Linear(10, 1)

    def forward(self, x):
        return torch.sigmoid(self.linear(x))

    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters(), lr=0.05)

    def training_step(self, batch, batch_idx):
        probs = self.forward(batch[0])
        label = batch[1].float()

        loss = self.criterion(probs, label)
        return {"loss": loss, "log": {"train_loss": loss}}

    def train_dataloader(self):
        return DataLoader(self.dataset, batch_size=5)


def test_lr(**kwargs):
    print(f"Testing for {kwargs}")

    X = torch.randn(100, 10)
    true_w = torch.randn(10, 1)
    y = torch.sigmoid(torch.matmul(X, true_w)) > 0.5

    dataset = TensorDataset(X, y)
    lr = LogisticRegressionClassifier(dataset)
    trainer = pl.Trainer(max_epochs=5, default_root_dir="/tmp/", **kwargs)
    abs_sum = torch.sum(torch.abs(lr.linear.weight))
    print(lr.linear.weight)
    trainer.fit(lr)
    abs_sum2 = torch.sum(torch.abs(lr.linear.weight))
    print(lr.linear.weight)
    assert torch.abs(abs_sum - abs_sum2).item() > 0.1, "Training should happen"


if __name__ == "__main__":
    import os
    import random

    port = random.randint(12000, 19000)
    print("Port=", port)
    os.environ["MASTER_PORT"] = str(port)
    os.environ["MASTER_ADDR"] = "127.0.0.1"

    test_lr()

    test_lr(gpus=1)

    test_lr(gpus=2)
```

**Comment by williamFalcon** [comment]
1. Lightning sets the port and address for you... you don't need to do it.
2. it hangs because you use the same port 3 times... and likely the processes are still going

```python
#!/usr/bin/env python3
import pytorch_lightning as pl
import torch
from torch import nn
from torch.utils.data import DataLoader, TensorDataset


class LogisticRegressionClassifier(pl.LightningModule):
    def __init__(self, dataset):
        super().__init__()
        self.criterion = nn.BCELoss()
        self.dataset = dataset
        self.linear = nn.Linear(10, 1)

    def forward(self, x):
        return torch.sigmoid(self.linear(x))

    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters(), lr=0.05)

    def training_step(self, batch, batch_idx):
        probs = self.forward(batch[0])
        label = batch[1].float()

        loss = self.criterion(probs, label)
        return {"loss": loss, "log": {"train_loss": loss}}

    def train_dataloader(self):
        return DataLoader(self.dataset, batch_size=5)


def test_lr(**kwargs):
    print(f"Testing for {kwargs}")

    X = torch.randn(100, 10)
    true_w = torch.randn(10, 1)
    y = torch.sigmoid(torch.matmul(X, true_w)) > 0.5

    dataset = TensorDataset(X, y)
    lr = LogisticRegressionClassifier(dataset)
    trainer = pl.Trainer(max_epochs=5, default_root_dir="/tmp/", **kwargs)
    abs_sum = torch.sum(torch.abs(lr.linear.weight))
    print(lr.linear.weight)
    trainer.fit(lr)
    abs_sum2 = torch.sum(torch.abs(lr.linear.weight))
    print(lr.linear.weight)
    assert torch.abs(abs_sum - abs_sum2).item() > 0.1, "Training should happen"


if __name__ == "__main__":
    import os
    import random

    test_lr()
```

2. The new ddp assumes that you are running a single file once... so, you can only use one of these test_lr functions.

3. old ddp had a lot of problems but this one has fewer problems, however requires running as a script. so, we can add support for the old way, but it's not going to work well.

**Comment by williamFalcon** [comment]
image

File: tests/base/develop_pipelines.py [hunk 1]
```diff
@@ -47,10 +47,14 @@ def run_model_test(trainer_options, model, on_gpu: bool = True, version=None, wi
         trainer_options.update(checkpoint_callback=True)
 
     trainer = Trainer(**trainer_options)
+    initial_values = torch.tensor([torch.sum(torch.abs(x)) for x in model.parameters()])
     result = trainer.fit(model)
+    post_train_values = torch.tensor([torch.sum(torch.abs(x)) for x in model.parameters()])
 
     # correct result and ok accuracy
     assert result == 1, 'trainer failed'
```
Comment (awaelchli) [inline tests/base/develop_pipelines.py:55]: can you remove the comment? the assertion does not actually do what the comment says :)

File: tests/base/develop_pipelines.py [hunk 2]
```diff
@@ -47,10 +47,14 @@ def run_model_test(trainer_options, model, on_gpu: bool = True, version=None, wi
...
+    # Check that the model is actually changed post-training
+    assert torch.norm(initial_values - post_train_values) > 0.1
```
Comment (awaelchli) [inline tests/base/develop_pipelines.py:57]: Since we are working with discrete values here, I conjecture the probability that the model changes and lands at the same weights at the end of training is strictly > 0 😄

File: tests/base/develop_pipelines.py [hunk 3]
```diff
@@ -47,10 +47,14 @@ def run_model_test(trainer_options, model, on_gpu: bool = True, version=None, wi
         trainer_options.update(checkpoint_callback=True)
 
     trainer = Trainer(**trainer_options)
+    initial_values = torch.tensor([torch.sum(torch.abs(x)) for x in model.parameters()])
     result = trainer.fit(model)
+    post_train_values = torch.tensor([torch.sum(torch.abs(x)) for x in model.parameters()])
 
     # correct result and ok accuracy
```
Comment (awaelchli) [inline tests/base/develop_pipelines.py:54]:
```suggestion
```

**Review by awaelchli — APPROVED** [review]

**Review by justusschock — APPROVED** [review]
